pkg/server/plugin/upstreamca/awssecret/awslib.go

```diff
@@ -0,0 +1,50 @@
+package awssecret
+
+import (
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/request"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/iam"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/aws/aws-sdk-go/service/secretsmanager/secretsmanageriface"
+)
+
+type awsClient interface {
+	DescribeInstancesWithContext(aws.Context, *ec2.DescribeInstancesInput, ...request.Option) (*ec2.DescribeInstancesOutput, error)
+	GetInstanceProfileWithContext(aws.Context, *iam.GetInstanceProfileInput, ...request.Option) (*iam.GetInstanceProfileOutput, error)
+}
+
+func readARN(sm secretsmanageriface.SecretsManagerAPI, arn string) (*string, error) {
```

Comment:
I wanted to enable returning `nil` in the error case. Wouldn't it be more confusing to return an empty string?

Reply:
Returning (string, error) is idiomatic in go when you don't need to disambiguate between "unset" and "empty" (for a fun exercise grep for `(string, error)` in the golang codebase). In this particular case, an empty ARN would never make sense, so it stands that either your function is going to return ("", some error) or ("somearn", nil). Making callers dereference the pointer is just added friction.

Reply:
In addition, having this function ensure that resp.SecretString is a valid pointer to a non-empty string (and failing the call otherwise) removes that burden from the caller.